// @stackframe/react/dist/esm/components-page/account-settings.js
"use client";
"use client";

// src/components-page/account-settings.tsx
import { Skeleton, Typography } from "@stackframe/stack-ui";
import { Contact, ShieldCheck, Bell, Monitor, Key, Settings, CirclePlus } from "lucide-react";
import { Suspense } from "react";
import { useStackApp, useUser } from "../index.js";
import { MaybeFullPage } from "../components/elements/maybe-full-page.js";
import { SidebarLayout } from "../components/elements/sidebar-layout.js";
import { TeamIcon } from "../components/team-icon.js";
import { useTranslation } from "../lib/translations.js";
import { ActiveSessionsPage } from "./account-settings/active-sessions/active-sessions-page.js";
import { ApiKeysPage } from "./account-settings/api-keys/api-keys-page.js";
import { EmailsAndAuthPage } from "./account-settings/email-and-auth/email-and-auth-page.js";
import { NotificationsPage } from "./account-settings/notifications/notifications-page.js";
import { ProfilePage } from "./account-settings/profile-page/profile-page.js";
import { SettingsPage } from "./account-settings/settings/settings-page.js";
import { TeamCreationPage } from "./account-settings/teams/team-creation-page.js";
import { TeamPage } from "./account-settings/teams/team-page.js";
import { jsx, jsxs } from "react/jsx-runtime";
var iconMap = {
  Contact,
  ShieldCheck,
  Bell,
  Monitor,
  Key,
  Settings,
  CirclePlus
};
var Icon = ({ name }) => {
  const LucideIcon = iconMap[name];
  return /* @__PURE__ */ jsx(LucideIcon, { className: "mr-2 h-4 w-4" });
};
function AccountSettings(props) {
  const { t } = useTranslation();
  const userFromHook = useUser({ or: props.mockUser ? "return-null" : "redirect" });
  const stackApp = useStackApp();
  const projectFromHook = stackApp.useProject();
  const user = props.mockUser ? {
    useTeams: () => []
    // Mock empty teams for now
  } : userFromHook;
  const project = props.mockProject || projectFromHook;
  const teams = user?.useTeams() || [];
  if (!props.mockUser && !userFromHook) {
    return null;
  }
  return /* @__PURE__ */ jsx(MaybeFullPage, { fullPage: !!props.fullPage, children: /* @__PURE__ */ jsx("div", { className: "self-stretch flex-grow w-full", children: /* @__PURE__ */ jsx(
    SidebarLayout,
    {
      items: [
        {
          title: t("My Profile"),
          type: "item",
          id: "profile",
          icon: /* @__PURE__ */ jsx(Icon, { name: "Contact" }),
          content: /* @__PURE__ */ jsx(ProfilePage, { mockUser: props.mockUser })
        },
        {
          title: t("Emails & Auth"),
          type: "item",
          id: "auth",
          icon: /* @__PURE__ */ jsx(Icon, { name: "ShieldCheck" }),
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(EmailsAndAuthPageSkeleton, {}), children: /* @__PURE__ */ jsx(EmailsAndAuthPage, { mockMode: !!props.mockUser }) })
        },
        {
          title: t("Notifications"),
          type: "item",
          id: "notifications",
          icon: /* @__PURE__ */ jsx(Icon, { name: "Bell" }),
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(NotificationsPageSkeleton, {}), children: /* @__PURE__ */ jsx(NotificationsPage, {}) })
        },
        {
          title: t("Active Sessions"),
          type: "item",
          id: "sessions",
          icon: /* @__PURE__ */ jsx(Icon, { name: "Monitor" }),
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(ActiveSessionsPageSkeleton, {}), children: /* @__PURE__ */ jsx(ActiveSessionsPage, { mockSessions: props.mockSessions, mockMode: !!props.mockUser }) })
        },
        ...project.config.allowUserApiKeys ? [{
          title: t("API Keys"),
          type: "item",
          id: "api-keys",
          icon: /* @__PURE__ */ jsx(Icon, { name: "Key" }),
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(ApiKeysPageSkeleton, {}), children: /* @__PURE__ */ jsx(ApiKeysPage, { mockApiKeys: props.mockApiKeys, mockMode: !!props.mockUser }) })
        }] : [],
        {
          title: t("Settings"),
          type: "item",
          id: "settings",
          icon: /* @__PURE__ */ jsx(Icon, { name: "Settings" }),
          content: /* @__PURE__ */ jsx(SettingsPage, { mockMode: !!props.mockUser })
        },
        ...props.extraItems?.map((item) => ({
          title: item.title,
          type: "item",
          id: item.id,
          icon: (() => {
            const iconName = item.iconName;
            if (iconName) {
              return /* @__PURE__ */ jsx(Icon, { name: iconName });
            } else if (item.icon) {
              return item.icon;
            }
            return null;
          })(),
          content: item.content
        })) || [],
        ...teams.length > 0 || project.config.clientTeamCreationEnabled ? [{
          title: t("Teams"),
          type: "divider"
        }] : [],
        ...teams.map((team) => ({
          title: /* @__PURE__ */ jsxs("div", { className: "flex gap-2 items-center w-full", children: [
            /* @__PURE__ */ jsx(TeamIcon, { team }),
            /* @__PURE__ */ jsx(Typography, { className: "max-w-[320px] md:w-[90%] truncate", children: team.displayName })
          ] }),
          type: "item",
          id: `team-${team.id}`,
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(TeamPageSkeleton, {}), children: /* @__PURE__ */ jsx(TeamPage, { team }) })
        })),
        ...project.config.clientTeamCreationEnabled ? [{
          title: t("Create a team"),
          icon: /* @__PURE__ */ jsx(Icon, { name: "CirclePlus" }),
          type: "item",
          id: "team-creation",
          content: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(TeamCreationSkeleton, {}), children: /* @__PURE__ */ jsx(TeamCreationPage, { mockMode: !!props.mockUser }) })
        }] : []
      ].filter((p) => p.type === "divider" || p.content),
      title: t("Account Settings")
    }
  ) }) });
}
function PageLayout(props) {
  return /* @__PURE__ */ jsx("div", { className: "flex flex-col gap-6", children: props.children });
}
function EmailsAndAuthPageSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" })
  ] });
}
function ActiveSessionsPageSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-6 w-48 mb-2" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-4 w-full mb-4" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-[200px] w-full mt-1 rounded-md" })
  ] });
}
function ApiKeysPageSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-[200px] w-full mt-1 rounded-md" })
  ] });
}
function TeamPageSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-[200px] w-full mt-1 rounded-md" })
  ] });
}
function TeamCreationSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" })
  ] });
}
function NotificationsPageSkeleton() {
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" }),
    /* @__PURE__ */ jsx(Skeleton, { className: "h-9 w-full mt-1" })
  ] });
}
export {
  AccountSettings
};
