// @stackframe/react/dist/lib/cookie.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/cookie.ts
var cookie_exports = {};
__export(cookie_exports, {
  consumeVerifierAndStateCookie: () => consumeVerifierAndStateCookie,
  createBrowserCookieHelper: () => createBrowserCookieHelper,
  createCookieHelper: () => createCookieHelper,
  createPlaceholderCookieHelper: () => createPlaceholderCookieHelper,
  deleteCookie: () => deleteCookie,
  deleteCookieClient: () => deleteCookieClient,
  getAllCookiesClient: () => getAllCookiesClient,
  getCookie: () => getCookie,
  getCookieClient: () => getCookieClient,
  isSecure: () => isSecure,
  saveVerifierAndState: () => saveVerifierAndState,
  setCookie: () => setCookie,
  setCookieClient: () => setCookieClient,
  setOrDeleteCookie: () => setOrDeleteCookie,
  setOrDeleteCookieClient: () => setOrDeleteCookieClient
});
module.exports = __toCommonJS(cookie_exports);
var import_env = require("@stackframe/stack-shared/dist/utils/env");
var import_errors = require("@stackframe/stack-shared/dist/utils/errors");
var import_js_cookie = __toESM(require("js-cookie"));
var import_oauth4webapi = require("oauth4webapi");
function ensureClient() {
  if (!(0, import_env.isBrowserLike)()) {
    throw new Error("cookieClient functions can only be called in a browser environment, yet window is undefined");
  }
}
async function createPlaceholderCookieHelper() {
  function throwError() {
    throw new import_errors.StackAssertionError("Throwing cookie helper is just a placeholder. This should never be called");
  }
  return {
    get: throwError,
    getAll: throwError,
    set: throwError,
    setOrDelete: throwError,
    delete: throwError
  };
}
async function createCookieHelper() {
  if ((0, import_env.isBrowserLike)()) {
    return createBrowserCookieHelper();
  } else {
    return await createPlaceholderCookieHelper();
  }
}
function createBrowserCookieHelper() {
  return {
    get: getCookieClient,
    getAll: getAllCookiesClient,
    set: setCookieClient,
    setOrDelete: setOrDeleteCookieClient,
    delete: deleteCookieClient
  };
}
function getCookieClient(name) {
  const all = getAllCookiesClient();
  return all[name] ?? null;
}
function getAllCookiesClient() {
  ensureClient();
  import_js_cookie.default.set("stack-is-https", "true", { secure: true });
  return import_js_cookie.default.get();
}
async function getCookie(name) {
  const cookieHelper = await createCookieHelper();
  return cookieHelper.get(name);
}
async function isSecure() {
  if ((0, import_env.isBrowserLike)()) {
    return determineSecureFromClientContext();
  }
  return false;
}
function determineSecureFromClientContext() {
  return typeof window !== "undefined" && window.location.protocol === "https:";
}
function setCookieClientInternal(name, value, options = {}) {
  const secure = options.secure ?? determineSecureFromClientContext();
  import_js_cookie.default.set(name, value, {
    expires: options.maxAge === void 0 ? void 0 : new Date(Date.now() + options.maxAge * 1e3),
    domain: options.domain,
    secure
  });
}
function deleteCookieClientInternal(name, options = {}) {
  if (options.domain !== void 0) {
    import_js_cookie.default.remove(name, { domain: options.domain, secure: determineSecureFromClientContext() });
  }
  import_js_cookie.default.remove(name, { secure: determineSecureFromClientContext() });
}
function setOrDeleteCookieClient(name, value, options = {}) {
  ensureClient();
  if (value === null) {
    deleteCookieClientInternal(name, options);
  } else {
    setCookieClientInternal(name, value, options);
  }
}
async function setOrDeleteCookie(name, value, options = {}) {
  const cookieHelper = await createCookieHelper();
  cookieHelper.setOrDelete(name, value, options);
}
function deleteCookieClient(name, options = {}) {
  ensureClient();
  deleteCookieClientInternal(name, options);
}
async function deleteCookie(name, options = {}) {
  const cookieHelper = await createCookieHelper();
  cookieHelper.delete(name, options);
}
function setCookieClient(name, value, options = {}) {
  ensureClient();
  setCookieClientInternal(name, value, options);
}
async function setCookie(name, value, options = {}) {
  const cookieHelper = await createCookieHelper();
  cookieHelper.set(name, value, options);
}
async function saveVerifierAndState() {
  const codeVerifier = (0, import_oauth4webapi.generateRandomCodeVerifier)();
  const codeChallenge = await (0, import_oauth4webapi.calculatePKCECodeChallenge)(codeVerifier);
  const state = (0, import_oauth4webapi.generateRandomState)();
  await setCookie("stack-oauth-outer-" + state, codeVerifier, { maxAge: 60 * 60 });
  return {
    codeChallenge,
    state
  };
}
function consumeVerifierAndStateCookie(state) {
  ensureClient();
  const cookieName = "stack-oauth-outer-" + state;
  const codeVerifier = getCookieClient(cookieName);
  if (!codeVerifier) {
    return null;
  }
  deleteCookieClient(cookieName);
  return {
    codeVerifier
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  consumeVerifierAndStateCookie,
  createBrowserCookieHelper,
  createCookieHelper,
  createPlaceholderCookieHelper,
  deleteCookie,
  deleteCookieClient,
  getAllCookiesClient,
  getCookie,
  getCookieClient,
  isSecure,
  saveVerifierAndState,
  setCookie,
  setCookieClient,
  setOrDeleteCookie,
  setOrDeleteCookieClient
});
