// @stackframe/stack-shared/dist/esm/interface/crud/config.js
// src/interface/crud/config.ts
import { createCrud } from "../../crud.js";
import * as schemaFields from "../../schema-fields.js";
import { yupObject } from "../../schema-fields.js";
var configOverrideCrudAdminReadSchema = yupObject({}).defined();
var configOverrideCrudAdminUpdateSchema = yupObject({
  config_override_string: schemaFields.yupString().optional()
}).defined();
var configOverrideCrud = createCrud({
  adminReadSchema: configOverrideCrudAdminReadSchema,
  adminUpdateSchema: configOverrideCrudAdminUpdateSchema,
  docs: {
    adminUpdate: {
      summary: "Update the config",
      description: "Update the config for a project and branch with an override",
      tags: ["Config"]
    }
  }
});
var configCrudAdminReadSchema = yupObject({
  config_string: schemaFields.yupString().defined()
}).defined();
var configCrud = createCrud({
  adminReadSchema: configCrudAdminReadSchema,
  docs: {
    adminRead: {
      summary: "Get the config",
      description: "Get the config for a project and branch",
      tags: ["Config"]
    }
  }
});
export {
  configCrud,
  configCrudAdminReadSchema,
  configOverrideCrud,
  configOverrideCrudAdminReadSchema,
  configOverrideCrudAdminUpdateSchema
};
