// @stackframe/stack-shared/dist/utils/crypto.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/crypto.tsx
var crypto_exports = {};
__export(crypto_exports, {
  decrypt: () => decrypt,
  encrypt: () => encrypt,
  generateRandomValues: () => generateRandomValues,
  generateSecureRandomString: () => generateSecureRandomString,
  hash: () => hash,
  iteratedHash: () => iteratedHash
});
module.exports = __toCommonJS(crypto_exports);
var import_bytes = require("./bytes.js");
var import_errors = require("./errors.js");
var import_globals = require("./globals.js");
var import_results = require("./results.js");
function generateRandomValues(array) {
  if (!import_globals.globalVar.crypto) {
    throw new import_errors.StackAssertionError("Crypto API is not available in this environment. Are you using an old browser?");
  }
  if (!import_globals.globalVar.crypto.getRandomValues) {
    throw new import_errors.StackAssertionError("crypto.getRandomValues is not available in this environment. Are you using an old browser?");
  }
  return import_globals.globalVar.crypto.getRandomValues(array);
}
function generateSecureRandomString(minBitsOfEntropy = 224) {
  const base32CharactersCount = Math.ceil(minBitsOfEntropy / 5);
  const bytesCount = Math.ceil(base32CharactersCount * 5 / 8);
  const randomBytes = generateRandomValues(new Uint8Array(bytesCount));
  const str = (0, import_bytes.encodeBase32)(randomBytes);
  return str.slice(str.length - base32CharactersCount).toLowerCase();
}
async function getDerivedSymmetricKey(purpose, secret, salt) {
  const originalSecretKey = await crypto.subtle.importKey("raw", typeof secret === "string" ? new TextEncoder().encode(secret) : secret, "HKDF", false, ["deriveKey"]);
  return await crypto.subtle.deriveKey(
    {
      name: "HKDF",
      salt,
      hash: "SHA-256",
      info: new TextEncoder().encode(JSON.stringify([
        "stack-crypto-helper-derived-symmetric-key",
        purpose,
        typeof secret === "string" ? "string-key" : "binary-key",
        (0, import_bytes.encodeBase64)(salt)
      ]))
    },
    originalSecretKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}
async function encrypt({ purpose, secret, value }) {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const derivedSecretKey = await getDerivedSymmetricKey(purpose, secret, salt);
  const cipher = await crypto.subtle.encrypt({
    name: "AES-GCM",
    iv
  }, derivedSecretKey, value);
  const version = [1, 0];
  return new Uint8Array([...version, ...salt, ...iv, ...new Uint8Array(cipher)]);
}
async function decrypt({ purpose, secret, cipher }) {
  const version = cipher.slice(0, 2);
  if (version[0] !== 1 || version[1] !== 0) throw new import_errors.StackAssertionError("Invalid ciphertext version in decrypt(...); expected 0x0100", { purpose });
  const salt = cipher.slice(2, 18);
  const iv = cipher.slice(18, 30);
  const cipherBytes = cipher.slice(30);
  const derivedSecretKey = await getDerivedSymmetricKey(purpose, secret, salt);
  try {
    const plaintext = await crypto.subtle.decrypt({
      name: "AES-GCM",
      iv
    }, derivedSecretKey, cipherBytes);
    return import_results.Result.ok(new Uint8Array(plaintext));
  } catch (e) {
    if (e instanceof DOMException && e.name === "OperationError") {
      return import_results.Result.error(new Error("Invalid ciphertext or secret when decrypting encrypted value", { cause: e }));
    }
    throw e;
  }
}
async function hash(options) {
  return await iteratedHash({ ...options, iterations: 1 });
}
async function iteratedHash(options) {
  const stringOrUint8ArrayToUint8Array = (value) => typeof value === "string" ? new TextEncoder().encode(value) : value;
  const stringOrUint8ArrayToBase64 = (value) => (0, import_bytes.encodeBase64)(stringOrUint8ArrayToUint8Array(value));
  const input = await crypto.subtle.importKey(
    "raw",
    stringOrUint8ArrayToUint8Array(options.value),
    "PBKDF2",
    false,
    ["deriveBits"]
  );
  return new Uint8Array(await crypto.subtle.deriveBits({
    name: "PBKDF2",
    salt: new TextEncoder().encode(JSON.stringify([
      "stack-crypto-helper-iterated-hash",
      options.purpose,
      stringOrUint8ArrayToBase64(options.salt ?? ""),
      stringOrUint8ArrayToBase64(options.extra ?? "")
    ])),
    iterations: options.iterations,
    hash: "SHA-256"
  }, input, 256));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  decrypt,
  encrypt,
  generateRandomValues,
  generateSecureRandomString,
  hash,
  iteratedHash
});
