// @stackframe/stack-shared/dist/utils/dates.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/dates.tsx
var dates_exports = {};
__export(dates_exports, {
  FAR_FUTURE_DATE: () => FAR_FUTURE_DATE,
  addInterval: () => addInterval,
  fromNow: () => fromNow,
  fromNowDetailed: () => fromNowDetailed,
  getInputDatetimeLocalString: () => getInputDatetimeLocalString,
  getIntervalsElapsed: () => getIntervalsElapsed,
  isWeekend: () => isWeekend,
  subtractInterval: () => subtractInterval
});
module.exports = __toCommonJS(dates_exports);
var import_schema_fields = require("../schema-fields.js");
var import_errors = require("./errors.js");
var import_math = require("./math.js");
function isWeekend(date) {
  return date.getDay() === 0 || date.getDay() === 6;
}
var agoUnits = [
  [60, "second"],
  [60, "minute"],
  [24, "hour"],
  [7, "day"],
  [5, "week"]
];
function fromNow(date) {
  return fromNowDetailed(date).result;
}
function fromNowDetailed(date) {
  if (!(date instanceof Date)) {
    throw new Error(`fromNow only accepts Date objects (received: ${date})`);
  }
  const now = /* @__PURE__ */ new Date();
  const elapsed = now.getTime() - date.getTime();
  let remainingInUnit = Math.abs(elapsed) / 1e3;
  if (remainingInUnit < 15) {
    return {
      result: "just now",
      secondsUntilChange: 15 - remainingInUnit
    };
  }
  let unitInSeconds = 1;
  for (const [nextUnitSize, unitName] of agoUnits) {
    const rounded = Math.round(remainingInUnit);
    if (rounded < nextUnitSize) {
      if (elapsed < 0) {
        return {
          result: `in ${rounded} ${unitName}${rounded === 1 ? "" : "s"}`,
          secondsUntilChange: (0, import_math.remainder)((remainingInUnit - rounded + 0.5) * unitInSeconds, unitInSeconds)
        };
      } else {
        return {
          result: `${rounded} ${unitName}${rounded === 1 ? "" : "s"} ago`,
          secondsUntilChange: (0, import_math.remainder)((rounded - remainingInUnit - 0.5) * unitInSeconds, unitInSeconds)
        };
      }
    }
    unitInSeconds *= nextUnitSize;
    remainingInUnit /= nextUnitSize;
  }
  return {
    result: date.toLocaleDateString("en-US", { year: "numeric", month: "short", day: "numeric" }),
    secondsUntilChange: Infinity
  };
}
function getInputDatetimeLocalString(date) {
  date = new Date(date);
  date.setMinutes(date.getMinutes() - date.getTimezoneOffset());
  return date.toISOString().slice(0, 19);
}
function applyInterval(date, times, interval) {
  if (!import_schema_fields.intervalSchema.isValidSync(interval)) {
    throw new import_errors.StackAssertionError(`Invalid interval`, { interval });
  }
  const [amount, unit] = interval;
  switch (unit) {
    case "millisecond": {
      date.setMilliseconds(date.getMilliseconds() + amount * times);
      break;
    }
    case "second": {
      date.setSeconds(date.getSeconds() + amount * times);
      break;
    }
    case "minute": {
      date.setMinutes(date.getMinutes() + amount * times);
      break;
    }
    case "hour": {
      date.setHours(date.getHours() + amount * times);
      break;
    }
    case "day": {
      date.setDate(date.getDate() + amount * times);
      break;
    }
    case "week": {
      date.setDate(date.getDate() + amount * times * 7);
      break;
    }
    case "month": {
      date.setMonth(date.getMonth() + amount * times);
      break;
    }
    case "year": {
      date.setFullYear(date.getFullYear() + amount * times);
      break;
    }
    default: {
      throw new import_errors.StackAssertionError(`Invalid interval despite schema validation`, { interval });
    }
  }
  return date;
}
function subtractInterval(date, interval) {
  return applyInterval(date, -1, interval);
}
function addInterval(date, interval) {
  return applyInterval(date, 1, interval);
}
var FAR_FUTURE_DATE = /* @__PURE__ */ new Date(864e13);
function getMsPerDayIntervalUnit(unit) {
  if (unit === "day") {
    return 24 * 60 * 60 * 1e3;
  }
  return 7 * 24 * 60 * 60 * 1e3;
}
function getIntervalsElapsed(anchor, to, repeat) {
  const [amount, unit] = repeat;
  if (to <= anchor) return 0;
  if (unit === "day" || unit === "week") {
    const msPerUnit = getMsPerDayIntervalUnit(unit);
    const diffMs = to.getTime() - anchor.getTime();
    return Math.floor(diffMs / (msPerUnit * amount));
  }
  if (["month", "year"].includes(unit)) {
    let count = 0;
    let current = new Date(anchor);
    for (; ; ) {
      const next = addInterval(new Date(current), [amount, unit]);
      if (next > to) break;
      current = next;
      count += 1;
    }
    return count;
  }
  return 0;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FAR_FUTURE_DATE,
  addInterval,
  fromNow,
  fromNowDetailed,
  getInputDatetimeLocalString,
  getIntervalsElapsed,
  isWeekend,
  subtractInterval
});
