// @stackframe/react/dist/esm/components-page/account-settings/settings/sign-out-section.js
// src/components-page/account-settings/settings/sign-out-section.tsx
import { Button } from "@stackframe/stack-ui";
import { useUser } from "../../../lib/hooks.js";
import { useTranslation } from "../../../lib/translations.js";
import { Section } from "../section.js";
import { jsx } from "react/jsx-runtime";
function SignOutSection(props) {
  const { t } = useTranslation();
  const user = useUser({ or: props?.mockMode ? "return-null" : "throw" });
  const handleSignOut = async () => {
    if (props?.mockMode) {
      alert("Mock mode: Sign out clicked");
      return;
    }
    if (user) {
      await user.signOut();
    }
  };
  return /* @__PURE__ */ jsx(
    Section,
    {
      title: t("Sign out"),
      description: t("End your current session"),
      children: /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(
        Button,
        {
          variant: "secondary",
          onClick: handleSignOut,
          children: t("Sign out")
        }
      ) })
    }
  );
}
export {
  SignOutSection
};
