// @stackframe/react/dist/components/profile-image-editor.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/profile-image-editor.tsx
var profile_image_editor_exports = {};
__export(profile_image_editor_exports, {
  ProfileImageEditor: () => ProfileImageEditor,
  checkImageUrl: () => checkImageUrl,
  getCroppedImg: () => getCroppedImg
});
module.exports = __toCommonJS(profile_image_editor_exports);
var import_base64 = require("@stackframe/stack-shared/dist/utils/base64");
var import_promises = require("@stackframe/stack-shared/dist/utils/promises");
var import_stack_ui = require("@stackframe/stack-ui");
var import_browser_image_compression = __toESM(require("browser-image-compression"));
var import_lucide_react = require("lucide-react");
var import_react = require("react");
var import_react_easy_crop = __toESM(require("react-easy-crop"));
var import_translations = require("../lib/translations.js");
var import_user_avatar = require("./elements/user-avatar.js");
var import_jsx_runtime = require("react/jsx-runtime");
async function checkImageUrl(url) {
  try {
    const res = await fetch(url, { method: "HEAD" });
    const buff = await res.blob();
    return buff.type.startsWith("image/");
  } catch (e) {
    return false;
  }
}
var createImage = (url) => new Promise((resolve, reject) => {
  const image = new Image();
  image.addEventListener("load", () => resolve(image));
  image.addEventListener("error", (error) => reject(error));
  image.setAttribute("crossOrigin", "anonymous");
  image.src = url;
});
async function getCroppedImg(imageSrc, pixelCrop) {
  const image = await createImage(imageSrc);
  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    return null;
  }
  const safeCrop = {
    x: Math.max(0, pixelCrop.x),
    y: Math.max(0, pixelCrop.y),
    width: Math.max(1, pixelCrop.width),
    height: Math.max(1, pixelCrop.height)
  };
  canvas.width = safeCrop.width;
  canvas.height = safeCrop.height;
  ctx.drawImage(
    image,
    safeCrop.x,
    safeCrop.y,
    safeCrop.width,
    safeCrop.height,
    0,
    0,
    safeCrop.width,
    safeCrop.height
  );
  return canvas.toDataURL("image/jpeg");
}
function ProfileImageEditor(props) {
  const { t } = (0, import_translations.useTranslation)();
  const [rawUrl, setRawUrl] = (0, import_react.useState)(null);
  const [error, setError] = (0, import_react.useState)(null);
  const [crop, setCrop] = (0, import_react.useState)({ x: 0, y: 0 });
  const [zoom, setZoom] = (0, import_react.useState)(1);
  const [croppedAreaPixels, setCroppedAreaPixels] = (0, import_react.useState)(null);
  function reset() {
    setRawUrl(null);
    setError(null);
    setCrop({ x: 0, y: 0 });
    setZoom(1);
    setCroppedAreaPixels(null);
  }
  const onCropChange = (0, import_react.useCallback)((crop2) => {
    setCrop(crop2);
  }, []);
  const onCropComplete = (0, import_react.useCallback)((croppedArea, croppedAreaPixels2) => {
    setCroppedAreaPixels(croppedAreaPixels2);
  }, []);
  const onZoomChange = (0, import_react.useCallback)((zoom2) => {
    setZoom(zoom2);
  }, []);
  function upload() {
    const input = document.createElement("input");
    input.type = "file";
    input.onchange = (e) => {
      const file = e.target.files?.[0];
      if (!file) return;
      (0, import_promises.runAsynchronouslyWithAlert)(async () => {
        const rawUrl2 = await (0, import_base64.fileToBase64)(file);
        if (await checkImageUrl(rawUrl2)) {
          setRawUrl(rawUrl2);
          setError(null);
        } else {
          setError(t("Invalid image"));
        }
        input.remove();
      });
    };
    input.click();
  }
  if (!rawUrl) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex flex-col", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "cursor-pointer relative", onClick: upload, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_user_avatar.UserAvatar,
          {
            size: 60,
            user: props.user,
            border: true
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "absolute top-0 left-0 h-[60px] w-[60px] bg-gray-500/20 backdrop-blur-sm items-center justify-center rounded-full flex opacity-0 hover:opacity-100 transition-opacity", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "bg-background p-2 rounded-full", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_lucide_react.Upload, { className: "h-5 w-5" }) }) })
      ] }),
      error && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.Typography, { variant: "destructive", type: "label", children: error })
    ] });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex flex-col items-center gap-4", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "relative w-64 h-64", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_easy_crop.default,
      {
        image: rawUrl || props.user.profileImageUrl || "",
        crop,
        zoom,
        aspect: 1,
        cropShape: "round",
        showGrid: false,
        onCropChange,
        onCropComplete,
        onZoomChange
      }
    ) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_stack_ui.Slider,
      {
        min: 1,
        max: 3,
        step: 0.1,
        value: [zoom],
        onValueChange: (v) => onZoomChange(v[0])
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex flex-row gap-2", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_stack_ui.Button,
        {
          onClick: async () => {
            if (rawUrl && croppedAreaPixels) {
              const croppedImageUrl = await getCroppedImg(rawUrl, croppedAreaPixels);
              if (croppedImageUrl) {
                const compressedFile = await (0, import_browser_image_compression.default)(
                  await import_browser_image_compression.default.getFilefromDataUrl(croppedImageUrl, "profile-image"),
                  {
                    maxSizeMB: 0.1,
                    fileType: "image/jpeg"
                  }
                );
                const compressedUrl = await import_browser_image_compression.default.getDataUrlFromFile(compressedFile);
                await props.onProfileImageUrlChange(compressedUrl);
                reset();
              } else {
                setError(t("Could not crop image."));
              }
            }
          },
          children: t("Save")
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_stack_ui.Button,
        {
          variant: "secondary",
          onClick: reset,
          children: t("Cancel")
        }
      )
    ] })
  ] });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ProfileImageEditor,
  checkImageUrl,
  getCroppedImg
});
