// @stackframe/react/dist/components/team-switcher.js
"use client";
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/team-switcher.tsx
var team_switcher_exports = {};
__export(team_switcher_exports, {
  TeamSwitcher: () => TeamSwitcher
});
module.exports = __toCommonJS(team_switcher_exports);
var import_errors = require("@stackframe/stack-shared/dist/utils/errors");
var import_promises = require("@stackframe/stack-shared/dist/utils/promises");
var import_stack_ui = require("@stackframe/stack-ui");
var import_lucide_react = require("lucide-react");
var import_react = require("react");
var import__ = require("../index.js");
var import_translations = require("../lib/translations.js");
var import_team_icon = require("./team-icon.js");
var import_jsx_runtime = require("react/jsx-runtime");
function TeamSwitcher(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Suspense, { fallback: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Fallback, {}), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Inner, { ...props }) });
}
function Fallback() {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.Skeleton, { className: "h-9 w-full max-w-64 stack-scope" });
}
function Inner(props) {
  const { t } = (0, import_translations.useTranslation)();
  const appFromHook = (0, import__.useStackApp)();
  const userFromHook = (0, import__.useUser)();
  const app = props.mockUser ? {
    useProject: () => props.mockProject || { config: { clientTeamCreationEnabled: false } },
    useNavigate: () => () => {
    },
    // Mock navigate function
    urls: { accountSettings: "/account-settings" }
  } : appFromHook;
  const user = props.mockUser ? {
    selectedTeam: props.mockUser.team,
    useTeams: () => props.mockTeams || [],
    setSelectedTeam: async () => {
    }
    // Mock function
  } : userFromHook;
  const navigate = app.useNavigate();
  const project = app.useProject();
  const rawTeams = user?.useTeams();
  const selectedTeam = props.team || rawTeams?.find((team) => team.id === props.teamId);
  const teams = (0, import_react.useMemo)(() => rawTeams?.sort((a, b) => b.id === selectedTeam?.id ? 1 : -1), [rawTeams, selectedTeam]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_stack_ui.Select,
    {
      value: selectedTeam?.id || (props.allowNull ? "null-sentinel" : void 0),
      onValueChange: (value) => {
        (0, import_promises.runAsynchronouslyWithAlert)(async () => {
          let team = null;
          if (value !== "null-sentinel") {
            team = teams?.find((team2) => team2.id === value) || null;
            if (!team) {
              throw new import_errors.StackAssertionError("Team not found, this should not happen");
            }
          } else {
            team = null;
          }
          if (props.onChange) {
            await props.onChange(team);
          }
        });
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectTrigger, { className: (0, import_stack_ui.cn)("stack-scope max-w-64", props.triggerClassName), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectValue, { placeholder: "Select team" }) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_stack_ui.SelectContent, { className: "stack-scope", children: [
          selectedTeam ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_stack_ui.SelectGroup, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectLabel, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex items-center justify-between", children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: t("Current team") }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_stack_ui.Button,
                {
                  variant: "ghost",
                  size: "icon",
                  className: "h-6 w-6",
                  onClick: () => {
                    if (!props.mockUser) {
                      navigate(`${app.urls.accountSettings}#team-${selectedTeam.id}`);
                    }
                  },
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_lucide_react.Settings, { className: "h-4 w-4" })
                }
              )
            ] }) }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectItem, { value: selectedTeam.id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex items-center gap-2", children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_team_icon.TeamIcon, { team: selectedTeam }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.Typography, { className: "max-w-40 truncate", children: selectedTeam.displayName })
            ] }) })
          ] }) : void 0,
          props.allowNull && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectGroup, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectItem, { value: "null-sentinel", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex items-center gap-2", children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_team_icon.TeamIcon, { team: "personal" }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.Typography, { className: "max-w-40 truncate", children: props.nullLabel || t("No team") })
          ] }) }) }),
          teams?.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_stack_ui.SelectGroup, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectLabel, { children: t("Other teams") }),
            teams.filter((team) => team.id !== selectedTeam?.id).map((team) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectItem, { value: team.id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "flex items-center gap-2", children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_team_icon.TeamIcon, { team }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.Typography, { className: "max-w-64 truncate", children: team.displayName })
            ] }) }, team.id))
          ] }) : null,
          !teams?.length && !props.allowNull ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectGroup, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectLabel, { children: t("No teams yet") }) }) : null,
          project.config.clientTeamCreationEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_stack_ui.SelectSeparator, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_stack_ui.Button,
              {
                onClick: () => {
                  if (!props.mockUser) {
                    navigate(`${app.urls.accountSettings}#team-creation`);
                  }
                },
                className: "w-full",
                variant: "ghost",
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_lucide_react.PlusCircle, { className: "mr-2 h-4 w-4" }),
                  " ",
                  t("Create a team")
                ]
              }
            ) })
          ] })
        ] })
      ]
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TeamSwitcher
});
