// @stackframe/stack-shared/dist/esm/apps/apps-ui.js
// src/apps/apps-ui.tsx
import { typedIncludes } from "../utils/arrays.js";
import { ALL_APPS } from "./apps-config.js";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var appSquareWidthExpression = "max(min(11vw,180px),80px)";
var appSquarePaddingExpression = "max(min(1vw,1.5rem),0.25rem)";
function AppIcon({
  appId,
  IconComponent,
  LogoComponent,
  className,
  disabled,
  style,
  cn
}) {
  const svgGradients = (gradients) => {
    return /* @__PURE__ */ jsx("svg", { width: "0", height: "0", children: Object.entries(gradients).map(([id, gradient]) => {
      return /* @__PURE__ */ jsx("linearGradient", { id, x1: "100%", y1: "100%", x2: "0%", y2: "0%", children: gradient.map((color, index) => {
        return /* @__PURE__ */ jsx("stop", { stopColor: color, offset: `${index * 100 / (gradient.length - 1)}%` }, index);
      }) }, id);
    }) });
  };
  const app = ALL_APPS[appId];
  return /* @__PURE__ */ jsxs(
    "div",
    {
      style,
      className: cn(
        "relative w-24 h-24 rounded-[24.154%] overflow-hidden select-none",
        !disabled && "bg-[linear-gradient(45deg,#dde,#fff)] dark:bg-[linear-gradient(45deg,#222,#666)]",
        disabled && "border-gray-400/70 border-dashed border-4",
        className
      ),
      children: [
        /* @__PURE__ */ jsx("div", { className: cn("w-full h-full isolate relative"), children: LogoComponent ? /* @__PURE__ */ jsx(
          "div",
          {
            className: "absolute inset-[20%] w-[60%] h-[60%] rounded-[24.154%] overflow-hidden flex items-center justify-center border",
            style: {
              opacity: disabled ? 0.6 : 1
            },
            children: /* @__PURE__ */ jsx(LogoComponent, {})
          }
        ) : /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx("svg", { width: "0", height: "0", children: svgGradients({
            "app-icon-gradient-light": ["#c0f", "#66f", "#4af"],
            "app-icon-gradient-dark": ["#3ec", "#9af", "#a5f"],
            "app-icon-gradient-light-expert": ["#f0c", "#f66", "#fa4"],
            "app-icon-gradient-dark-expert": ["#f0c", "#f66", "#fa4"],
            "app-icon-gradient-light-integration": ["#E5AB00", "#FFBA00", "#F8DF80"],
            "app-icon-gradient-dark-integration": ["#E5AB00", "#FFBA00", "#F8DF80"]
          }) }),
          /* @__PURE__ */ jsx(
            IconComponent,
            {
              opacity: disabled ? 0.75 : 1,
              className: cn(
                "inset-[20%] w-[60%] h-[60%] bg-clip-text text-transparent text-white absolute",
                typedIncludes(app.tags, "expert") ? "stroke-[url(#app-icon-gradient-light-expert)] dark:stroke-[url(#app-icon-gradient-dark-expert)]" : typedIncludes(app.tags, "integration") ? "stroke-[url(#app-icon-gradient-light-integration)] dark:stroke-[url(#app-icon-gradient-dark-integration)]" : "stroke-[url(#app-icon-gradient-light)] dark:stroke-[url(#app-icon-gradient-dark)]"
              )
            }
          )
        ] }) }),
        /* @__PURE__ */ jsx("div", { className: "absolute top-0 left-[-100%] right-0 flex flex-col gap-1 [transform:_rotate(-45deg)_translateY(24px)] [transform-origin:top_center]", children: app.stage !== "stable" && /* @__PURE__ */ jsx(
          "div",
          {
            className: cn(
              "h-4 uppercase text-xs font-bold font-mono tracking-widest text-center",
              disabled ? "bg-gray-400/80 text-white" : app.stage === "alpha" ? "bg-red-500 text-white" : "bg-yellow-600 text-white"
            ),
            children: app.stage
          }
        ) })
      ]
    }
  );
}
export {
  AppIcon,
  appSquarePaddingExpression,
  appSquareWidthExpression
};
