// @stackframe/stack-shared/dist/interface/crud/items.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/interface/crud/items.ts
var items_exports = {};
__export(items_exports, {
  itemCrud: () => itemCrud
});
module.exports = __toCommonJS(items_exports);
var import_crud = require("../../crud.js");
var import_schema_fields = require("../../schema-fields.js");
var itemReadSchema = (0, import_schema_fields.yupObject)({
  id: (0, import_schema_fields.yupString)().defined(),
  display_name: (0, import_schema_fields.yupString)().defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined()
}).defined();
var itemUpdateSchema = (0, import_schema_fields.yupObject)({
  delta: (0, import_schema_fields.yupNumber)().defined(),
  expires_at: (0, import_schema_fields.yupString)().optional(),
  description: (0, import_schema_fields.yupString)().optional(),
  allow_negative: (0, import_schema_fields.yupBoolean)().optional()
}).defined();
var itemCrud = (0, import_crud.createCrud)({
  clientReadSchema: itemReadSchema,
  serverUpdateSchema: itemUpdateSchema
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  itemCrud
});
