// @stackframe/stack-shared/dist/interface/crud/notification-preferences.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/interface/crud/notification-preferences.ts
var notification_preferences_exports = {};
__export(notification_preferences_exports, {
  notificationPreferenceCrud: () => notificationPreferenceCrud
});
module.exports = __toCommonJS(notification_preferences_exports);
var import_crud = require("../../crud.js");
var import_schema_fields = require("../../schema-fields.js");
var notificationPreferenceReadSchema = (0, import_schema_fields.yupObject)({
  notification_category_id: (0, import_schema_fields.yupString)().defined(),
  notification_category_name: (0, import_schema_fields.yupString)().defined(),
  enabled: (0, import_schema_fields.yupBoolean)().defined(),
  can_disable: (0, import_schema_fields.yupBoolean)().defined()
}).defined();
var notificationPreferenceUpdateSchema = (0, import_schema_fields.yupObject)({
  enabled: (0, import_schema_fields.yupBoolean)().defined()
}).defined();
var notificationPreferenceCrud = (0, import_crud.createCrud)({
  clientReadSchema: notificationPreferenceReadSchema,
  clientUpdateSchema: notificationPreferenceUpdateSchema
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  notificationPreferenceCrud
});
