// @stackframe/stack-shared/dist/interface/crud/products.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/interface/crud/products.ts
var products_exports = {};
__export(products_exports, {
  customerProductReadSchema: () => customerProductReadSchema,
  customerProductsListResponseSchema: () => customerProductsListResponseSchema
});
module.exports = __toCommonJS(products_exports);
var import_schema_fields = require("../../schema-fields.js");
var customerProductReadSchema = (0, import_schema_fields.yupObject)({
  id: (0, import_schema_fields.yupString)().nullable().defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined(),
  product: import_schema_fields.inlineProductSchema.defined()
}).defined();
var customerProductsListResponseSchema = (0, import_schema_fields.yupObject)({
  items: (0, import_schema_fields.yupArray)(customerProductReadSchema).defined(),
  is_paginated: (0, import_schema_fields.yupBoolean)().oneOf([true]).defined(),
  pagination: (0, import_schema_fields.yupObject)({
    next_cursor: (0, import_schema_fields.yupString)().nullable().defined()
  }).defined()
}).defined();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  customerProductReadSchema,
  customerProductsListResponseSchema
});
