// @stackframe/stack-shared/dist/interface/crud/transactions.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/interface/crud/transactions.ts
var transactions_exports = {};
__export(transactions_exports, {
  TRANSACTION_TYPES: () => TRANSACTION_TYPES,
  transactionEntrySchema: () => transactionEntrySchema,
  transactionSchema: () => transactionSchema
});
module.exports = __toCommonJS(transactions_exports);
var import_schema_fields = require("../../schema-fields.js");
var import_currency_constants = require("../../utils/currency-constants.js");
var import_objects = require("../../utils/objects.js");
var import_errors = require("../../utils/errors.js");
var USD_CURRENCY = import_currency_constants.SUPPORTED_CURRENCIES.find((currency) => currency.code === "USD") ?? (0, import_errors.throwErr)("USD currency configuration missing in SUPPORTED_CURRENCIES");
var chargedAmountSchema = (0, import_schema_fields.yupObject)({
  ...(0, import_objects.typedFromEntries)(import_currency_constants.SUPPORTED_CURRENCIES.map((currency) => [currency.code, (0, import_schema_fields.moneyAmountSchema)(currency).optional()]))
}).noUnknown(true).test("at-least-one-currency", "charged_amount must include at least one currency amount", (value) => {
  return Object.values(value).some((amount) => typeof amount === "string");
}).defined();
var netAmountSchema = (0, import_schema_fields.yupObject)({
  USD: (0, import_schema_fields.moneyAmountSchema)(USD_CURRENCY).defined()
}).noUnknown(true).defined();
var transactionEntryMoneyTransferSchema = (0, import_schema_fields.yupObject)({
  type: (0, import_schema_fields.yupString)().oneOf(["money_transfer"]).defined(),
  adjusted_transaction_id: (0, import_schema_fields.yupString)().nullable().defined(),
  adjusted_entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).nullable().defined(),
  customer_type: import_schema_fields.customerTypeSchema.defined(),
  customer_id: (0, import_schema_fields.yupString)().defined(),
  charged_amount: chargedAmountSchema,
  net_amount: netAmountSchema
}).defined();
var transactionEntryItemQuantityChangeSchema = (0, import_schema_fields.yupObject)({
  type: (0, import_schema_fields.yupString)().oneOf(["item_quantity_change"]).defined(),
  adjusted_transaction_id: (0, import_schema_fields.yupString)().nullable().defined(),
  adjusted_entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).nullable().defined(),
  customer_type: import_schema_fields.customerTypeSchema.defined(),
  customer_id: (0, import_schema_fields.yupString)().defined(),
  item_id: (0, import_schema_fields.yupString)().defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined()
}).defined();
var transactionEntryProductGrantSchema = (0, import_schema_fields.yupObject)({
  type: (0, import_schema_fields.yupString)().oneOf(["product_grant"]).defined(),
  adjusted_transaction_id: (0, import_schema_fields.yupString)().nullable().defined(),
  adjusted_entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).nullable().defined(),
  customer_type: import_schema_fields.customerTypeSchema.defined(),
  customer_id: (0, import_schema_fields.yupString)().defined(),
  product_id: (0, import_schema_fields.yupString)().nullable().defined(),
  product: import_schema_fields.inlineProductSchema.defined(),
  price_id: (0, import_schema_fields.yupString)().nullable().defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined(),
  subscription_id: (0, import_schema_fields.yupString)().optional(),
  one_time_purchase_id: (0, import_schema_fields.yupString)().optional()
}).test(
  "exclusive-reference",
  "subscription_id and one_time_purchase_id cannot both be set",
  (value, context) => {
    if (value.subscription_id != null && value.one_time_purchase_id != null) {
      return context.createError({
        message: "subscription_id and one_time_purchase_id cannot both be set"
      });
    }
    return true;
  }
).defined();
var transactionEntryProductRevocationSchema = (0, import_schema_fields.yupObject)({
  type: (0, import_schema_fields.yupString)().oneOf(["product_revocation"]).defined(),
  adjusted_transaction_id: (0, import_schema_fields.yupString)().defined(),
  adjusted_entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined()
}).defined();
var transactionEntryProductRevocationReversalSchema = (0, import_schema_fields.yupObject)({
  type: (0, import_schema_fields.yupString)().oneOf(["product_revocation_reversal"]).defined(),
  adjusted_transaction_id: (0, import_schema_fields.yupString)().defined(),
  adjusted_entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).defined(),
  quantity: (0, import_schema_fields.yupNumber)().defined()
}).defined();
var transactionEntrySchema = (0, import_schema_fields.yupUnion)(
  transactionEntryMoneyTransferSchema,
  transactionEntryItemQuantityChangeSchema,
  transactionEntryProductGrantSchema,
  transactionEntryProductRevocationSchema,
  transactionEntryProductRevocationReversalSchema
).defined();
var TRANSACTION_TYPES = [
  "purchase",
  "subscription-cancellation",
  "subscription-renewal",
  "manual-item-quantity-change",
  "chargeback",
  // todo
  "product-change"
  // todo
];
var transactionSchema = (0, import_schema_fields.yupObject)({
  id: (0, import_schema_fields.yupString)().defined(),
  created_at_millis: (0, import_schema_fields.yupNumber)().defined(),
  effective_at_millis: (0, import_schema_fields.yupNumber)().defined(),
  type: (0, import_schema_fields.yupString)().oneOf(TRANSACTION_TYPES).nullable().defined(),
  entries: (0, import_schema_fields.yupArray)(transactionEntrySchema).defined(),
  adjusted_by: (0, import_schema_fields.yupArray)(
    (0, import_schema_fields.yupObject)({
      transaction_id: (0, import_schema_fields.yupString)().defined(),
      entry_index: (0, import_schema_fields.yupNumber)().integer().min(0).defined()
    }).defined()
  ).defined(),
  test_mode: (0, import_schema_fields.yupBoolean)().defined()
}).defined();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TRANSACTION_TYPES,
  transactionEntrySchema,
  transactionSchema
});
