// @stackframe/react/dist/esm/components-page/account-settings/profile-page/profile-page.js
// src/components-page/account-settings/profile-page/profile-page.tsx
import { ProfileImageEditor } from "../../../components/profile-image-editor.js";
import { useUser } from "../../../lib/hooks.js";
import { useTranslation } from "../../../lib/translations.js";
import { EditableText } from "../editable-text.js";
import { PageLayout } from "../page-layout.js";
import { Section } from "../section.js";
import { jsx, jsxs } from "react/jsx-runtime";
function ProfilePage(props) {
  const { t } = useTranslation();
  const userFromHook = useUser({ or: props?.mockUser ? "return-null" : "redirect" });
  const user = props?.mockUser ? {
    displayName: props.mockUser.displayName || "John Doe",
    profileImageUrl: props.mockUser.profileImageUrl || null,
    update: async () => {
      console.log("Mock update called");
    }
  } : userFromHook;
  if (!user) {
    return null;
  }
  return /* @__PURE__ */ jsxs(PageLayout, { children: [
    /* @__PURE__ */ jsx(
      Section,
      {
        title: t("User name"),
        description: t("This is a display name and is not used for authentication"),
        children: /* @__PURE__ */ jsx(
          EditableText,
          {
            value: user.displayName || "",
            onSave: async (newDisplayName) => {
              await user.update({ displayName: newDisplayName });
            }
          }
        )
      }
    ),
    /* @__PURE__ */ jsx(
      Section,
      {
        title: t("Profile image"),
        description: t("Upload your own image as your avatar"),
        children: /* @__PURE__ */ jsx(
          ProfileImageEditor,
          {
            user,
            onProfileImageUrlChange: async (profileImageUrl) => {
              await user.update({ profileImageUrl });
            }
          }
        )
      }
    )
  ] });
}
export {
  ProfilePage
};
