// @stackframe/react/dist/esm/components/user-button.js
"use client";
"use client";

// src/components/user-button.tsx
import { runAsynchronouslyWithAlert } from "@stackframe/stack-shared/dist/utils/promises";
import { DropdownMenu, DropdownMenuContent, DropdownMenuItem, DropdownMenuLabel, DropdownMenuSeparator, DropdownMenuTrigger, Skeleton, Typography } from "@stackframe/stack-ui";
import { CircleUser, LogIn, LogOut, SunMoon, UserPlus } from "lucide-react";
import { Suspense } from "react";
import { useStackApp, useUser } from "../index.js";
import { useTranslation } from "../lib/translations.js";
import { UserAvatar } from "./elements/user-avatar.js";
import { jsx, jsxs } from "react/jsx-runtime";
function Item(props) {
  return /* @__PURE__ */ jsx(DropdownMenuItem, { onClick: () => runAsynchronouslyWithAlert(props.onClick), children: /* @__PURE__ */ jsxs("div", { className: "flex gap-2 items-center", children: [
    props.icon,
    /* @__PURE__ */ jsx(Typography, { children: props.text })
  ] }) });
}
function UserButton(props) {
  return /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(Skeleton, { className: "h-[34px] w-[34px] rounded-full stack-scope" }), children: /* @__PURE__ */ jsx(UserButtonInner, { ...props }) });
}
function UserButtonInner(props) {
  const userFromHook = useUser();
  const user = props.mockUser ? {
    displayName: props.mockUser.displayName || "Mock User",
    primaryEmail: props.mockUser.primaryEmail || "mock@example.com",
    profileImageUrl: props.mockUser.profileImageUrl,
    signOut: () => {
      console.log("Mock sign out - no action taken in demo mode");
      return Promise.resolve();
    }
  } : userFromHook;
  return /* @__PURE__ */ jsx(UserButtonInnerInner, { ...props, user });
}
function UserButtonInnerInner(props) {
  const { t } = useTranslation();
  const user = props.user;
  const app = useStackApp();
  const iconProps = { size: 20, className: "h-4 w-4" };
  return /* @__PURE__ */ jsxs(DropdownMenu, { children: [
    /* @__PURE__ */ jsx(DropdownMenuTrigger, { className: "outline-none stack-scope", children: /* @__PURE__ */ jsxs("div", { className: "flex gap-2 items-center", children: [
      /* @__PURE__ */ jsx(UserAvatar, { user }),
      user && props.showUserInfo && /* @__PURE__ */ jsxs("div", { className: "flex flex-col justify-center text-left", children: [
        /* @__PURE__ */ jsx("div", { className: "max-w-40 truncate", children: user.displayName }),
        /* @__PURE__ */ jsx("div", { className: "max-w-40 truncate text-sm text-gray-500", children: user.primaryEmail })
      ] })
    ] }) }),
    /* @__PURE__ */ jsxs(DropdownMenuContent, { className: "stack-scope", children: [
      /* @__PURE__ */ jsx(DropdownMenuLabel, { children: /* @__PURE__ */ jsxs("div", { className: "flex gap-2 items-center", children: [
        /* @__PURE__ */ jsx(UserAvatar, { user }),
        /* @__PURE__ */ jsxs("div", { children: [
          user && /* @__PURE__ */ jsx(Typography, { className: "max-w-40 truncate", children: user.displayName }),
          user && /* @__PURE__ */ jsx(Typography, { className: "max-w-40 truncate", variant: "secondary", type: "label", children: user.primaryEmail }),
          !user && /* @__PURE__ */ jsx(Typography, { children: t("Not signed in") })
        ] })
      ] }) }),
      /* @__PURE__ */ jsx(DropdownMenuSeparator, {}),
      user && /* @__PURE__ */ jsx(
        Item,
        {
          text: t("Account settings"),
          onClick: async () => {
            if (props.mockUser) {
              console.log("Mock account settings - no navigation in demo mode");
            } else {
              await app.redirectToAccountSettings();
            }
          },
          icon: /* @__PURE__ */ jsx(CircleUser, { ...iconProps })
        }
      ),
      !user && /* @__PURE__ */ jsx(
        Item,
        {
          text: t("Sign in"),
          onClick: async () => {
            if (props.mockUser) {
              console.log("Mock sign in - no navigation in demo mode");
            } else {
              await app.redirectToSignIn();
            }
          },
          icon: /* @__PURE__ */ jsx(LogIn, { ...iconProps })
        }
      ),
      !user && /* @__PURE__ */ jsx(
        Item,
        {
          text: t("Sign up"),
          onClick: async () => {
            if (props.mockUser) {
              console.log("Mock sign up - no navigation in demo mode");
            } else {
              await app.redirectToSignUp();
            }
          },
          icon: /* @__PURE__ */ jsx(UserPlus, { ...iconProps })
        }
      ),
      user && props.extraItems && props.extraItems.map((item, index) => /* @__PURE__ */ jsx(Item, { ...item }, index)),
      props.colorModeToggle && /* @__PURE__ */ jsx(
        Item,
        {
          text: t("Toggle theme"),
          onClick: props.colorModeToggle,
          icon: /* @__PURE__ */ jsx(SunMoon, { ...iconProps })
        }
      ),
      user && /* @__PURE__ */ jsx(
        Item,
        {
          text: t("Sign out"),
          onClick: async () => {
            if (props.mockUser) {
              console.log("Mock sign out - no action taken in demo mode");
            } else {
              await user.signOut();
            }
          },
          icon: /* @__PURE__ */ jsx(LogOut, { ...iconProps })
        }
      )
    ] })
  ] });
}
export {
  UserButton
};
