// @stackframe/react/dist/esm/index.js
// src/index.ts
export * from "./lib/stack-app/index.js";
import { getConvexProvidersConfig } from "./integrations/convex.js";
import { default as default2 } from "./components-page/stack-handler.js";
import { useStackApp, useUser } from "./lib/hooks.js";
import { default as default3 } from "./providers/stack-provider.js";
import { StackTheme } from "./providers/theme-provider.js";
import { AccountSettings } from "./components-page/account-settings.js";
import { AuthPage } from "./components-page/auth-page.js";
import { CliAuthConfirmation } from "./components-page/cli-auth-confirm.js";
import { EmailVerification } from "./components-page/email-verification.js";
import { ForgotPassword } from "./components-page/forgot-password.js";
import { PasswordReset } from "./components-page/password-reset.js";
import { SignIn } from "./components-page/sign-in.js";
import { SignUp } from "./components-page/sign-up.js";
import { CredentialSignIn } from "./components/credential-sign-in.js";
import { CredentialSignUp } from "./components/credential-sign-up.js";
import { UserAvatar } from "./components/elements/user-avatar.js";
import { MagicLinkSignIn } from "./components/magic-link-sign-in.js";
import { MessageCard } from "./components/message-cards/message-card.js";
import { OAuthButton } from "./components/oauth-button.js";
import { OAuthButtonGroup } from "./components/oauth-button-group.js";
import { SelectedTeamSwitcher } from "./components/selected-team-switcher.js";
import { TeamSwitcher } from "./components/team-switcher.js";
import { UserButton } from "./components/user-button.js";
export {
  AccountSettings,
  AuthPage,
  CliAuthConfirmation,
  CredentialSignIn,
  CredentialSignUp,
  EmailVerification,
  ForgotPassword,
  MagicLinkSignIn,
  MessageCard,
  OAuthButton,
  OAuthButtonGroup,
  PasswordReset,
  SelectedTeamSwitcher,
  SignIn,
  SignUp,
  default2 as StackHandler,
  default3 as StackProvider,
  StackTheme,
  TeamSwitcher,
  UserAvatar,
  UserButton,
  getConvexProvidersConfig,
  useStackApp,
  useUser
};
